In [1]:
import numpy as np

In [3]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
df = pd.read_csv("match_data.csv")

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3444: DtypeWarning: Columns (1) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 243817 entries, 0 to 243816
Data columns (total 23 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   match_id                243817 non-null  int64  
 1   season                  243817 non-null  object 
 2   start_date              243817 non-null  object 
 3   venue                   243817 non-null  object 
 4   innings                 243817 non-null  int64  
 5   ball                    243817 non-null  float64
 6   batting_team            243817 non-null  object 
 7   bowling_team            243817 non-null  object 
 8   striker                 243817 non-null  object 
 9   non_striker             243817 non-null  object 
 10  bowler                  243817 non-null  object 
 11  runs_off_bat            243817 non-null  int64  
 12  extras                  243817 non-null  int64  
 13  wides                   7652 non-null    float64
 14  noballs             

In [5]:
df.isna().sum()

id                       0
inning                   0
over                     0
ball                     0
batsman                  0
non_striker              0
bowler                   0
batsman_runs             0
extra_runs               0
total_runs               0
non_boundary             0
is_wicket                0
dismissal_kind      183973
player_dismissed    183973
fielder             186684
extras_type         183235
batting_team             0
bowling_team           191
dtype: int64

dismissal_kind, player_dismissed, fielder, extras_type have null values which is obvious. However, the bowling_team have 191 missing values which will be inspected.


In [9]:
df.head()

id  inning  over  ball      batsman  non_striker     bowler  \
0  335982       1     6     5   RT Ponting  BB McCullum  AA Noffke   
1  335982       1     6     6  BB McCullum   RT Ponting  AA Noffke   
2  335982       1     7     1  BB McCullum   RT Ponting     Z Khan   
3  335982       1     7     2  BB McCullum   RT Ponting     Z Khan   
4  335982       1     7     3   RT Ponting  BB McCullum     Z Khan   

   batsman_runs  extra_runs  total_runs  non_boundary  is_wicket  \
0             1           0           1             0          0   
1             1           0           1             0          0   
2             0           0           0             0          0   
3             1           0           1             0          0   
4             1           0           1             0          0   

  dismissal_kind player_dismissed fielder extras_type           batting_team  \
0            NaN              NaN     NaN         NaN  Kolkata Knight Riders   
1            NaN              NaN     NaN         NaN  Kolkata Knight Riders   
2            NaN              NaN     NaN         NaN  Kolkata Knight Riders   
3            NaN              NaN     NaN         NaN  Kolkata Knight Riders   
4            NaN              NaN     NaN         NaN  Kolkata Knight Riders   

                  bowling_team  
0  Royal Challengers Bangalore  
1  Royal Challengers Bangalore  
2  Royal Challengers Bangalore  
3  Royal Challengers Bangalore  
4  Royal Challengers Bangalore

In [6]:
df[df['bowling_team'].isna()]

id  inning  over  ball         batsman     non_striker  \
57048   501265       1     0     1         NV Ojha       DA Warner   
57049   501265       1     0     2       DA Warner         NV Ojha   
57050   501265       1     1     1       DA Warner         NV Ojha   
57051   501265       1     1     2         NV Ojha       DA Warner   
57052   501265       1     1     3         NV Ojha       DA Warner   
...        ...     ...   ...   ...             ...             ...   
115654  829763       1     7     3  AB de Villiers   Mandeep Singh   
115655  829763       1     8     6  AB de Villiers   Mandeep Singh   
115656  829763       1     9     1   Mandeep Singh  AB de Villiers   
115657  829763       1     8     4  AB de Villiers   Mandeep Singh   
115658  829763       1     8     5  AB de Villiers   Mandeep Singh   

             bowler  batsman_runs  extra_runs  total_runs  non_boundary  \
57048     AC Thomas             1           0           1             0   
57049     AC Thomas             0           1           1             0   
57050   JP Faulkner             0           1           1             0   
57051   JP Faulkner             0           0           0             0   
57052   JP Faulkner             0           0           0             0   
...             ...           ...         ...         ...           ...   
115654    STR Binny             0           0           0             0   
115655    SR Watson             2           0           2             0   
115656    STR Binny             4           0           4             0   
115657    SR Watson             0           0           0             0   
115658    SR Watson             0           0           0             0   

        is_wicket dismissal_kind player_dismissed fielder extras_type  \
57048           0            NaN              NaN     NaN         NaN   
57049           0            NaN              NaN     NaN     legbyes   
57050           0            NaN              NaN     NaN     legbyes   
57051           0            NaN              NaN     NaN         NaN   
57052           0            NaN              NaN     NaN         NaN   
...           ...            ...              ...     ...         ...   
115654          0            NaN              NaN     NaN         NaN   
115655          0            NaN              NaN     NaN         NaN   
115656          0            NaN              NaN     NaN         NaN   
115657          0            NaN              NaN     NaN         NaN   
115658          0            NaN              NaN     NaN         NaN   

                       batting_team bowling_team  
57048              Delhi Daredevils          NaN  
57049              Delhi Daredevils          NaN  
57050              Delhi Daredevils          NaN  
57051              Delhi Daredevils          NaN  
57052              Delhi Daredevils          NaN  
...                             ...          ...  
115654  Royal Challengers Bangalore          NaN  
115655  Royal Challengers Bangalore          NaN  
115656  Royal Challengers Bangalore          NaN  
115657  Royal Challengers Bangalore          NaN  
115658  Royal Challengers Bangalore          NaN  

[191 rows x 18 columns]

There is nothing that can be inferred from this. So, other csv file will be checked.

In [7]:
df2 = pd.read_csv("IPL Matches 2008-2020.csv")
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               816 non-null    int64  
 1   city             803 non-null    object 
 2   date             816 non-null    object 
 3   player_of_match  812 non-null    object 
 4   venue            816 non-null    object 
 5   neutral_venue    816 non-null    int64  
 6   team1            816 non-null    object 
 7   team2            816 non-null    object 
 8   toss_winner      816 non-null    object 
 9   toss_decision    816 non-null    object 
 10  winner           812 non-null    object 
 11  result           812 non-null    object 
 12  result_margin    799 non-null    float64
 13  eliminator       812 non-null    object 
 14  method           19 non-null     object 
 15  umpire1          816 non-null    object 
 16  umpire2          816 non-null    object 
dtypes: float64(1), i

In [8]:
df2.head()

id        city        date player_of_match  \
0  335982   Bangalore  2008-04-18     BB McCullum   
1  335983  Chandigarh  2008-04-19      MEK Hussey   
2  335984       Delhi  2008-04-19     MF Maharoof   
3  335985      Mumbai  2008-04-20      MV Boucher   
4  335986     Kolkata  2008-04-20       DJ Hussey   

                                        venue  neutral_venue  \
0                       M Chinnaswamy Stadium              0   
1  Punjab Cricket Association Stadium, Mohali              0   
2                            Feroz Shah Kotla              0   
3                            Wankhede Stadium              0   
4                                Eden Gardens              0   

                         team1                        team2  \
0  Royal Challengers Bangalore        Kolkata Knight Riders   
1              Kings XI Punjab          Chennai Super Kings   
2             Delhi Daredevils             Rajasthan Royals   
3               Mumbai Indians  Royal Challengers Bangalore   
4        Kolkata Knight Riders              Deccan Chargers   

                   toss_winner toss_decision                       winner  \
0  Royal Challengers Bangalore         field        Kolkata Knight Riders   
1          Chennai Super Kings           bat          Chennai Super Kings   
2             Rajasthan Royals           bat             Delhi Daredevils   
3               Mumbai Indians           bat  Royal Challengers Bangalore   
4              Deccan Chargers           bat        Kolkata Knight Riders   

    result  result_margin eliminator method    umpire1         umpire2  
0     runs          140.0          N    NaN  Asad Rauf     RE Koertzen  
1     runs           33.0          N    NaN  MR Benson      SL Shastri  
2  wickets            9.0          N    NaN  Aleem Dar  GA Pratapkumar  
3  wickets            5.0          N    NaN   SJ Davis       DJ Harper  
4  wickets            5.0          N    NaN  BF Bowden     K Hariharan

id is column is same in both the tables. It will be helpful in deciding the missing fields in bowling_team.

In [12]:
df3 = pd.merge(df, df2, on='id', how='left')
df3

id  inning  over  ball      batsman  non_striker  \
0        335982       1     6     5   RT Ponting  BB McCullum   
1        335982       1     6     6  BB McCullum   RT Ponting   
2        335982       1     7     1  BB McCullum   RT Ponting   
3        335982       1     7     2  BB McCullum   RT Ponting   
4        335982       1     7     3   RT Ponting  BB McCullum   
...         ...     ...   ...   ...          ...          ...   
193463  1237181       1    12     5      RR Pant      SS Iyer   
193464  1237181       1    12     6      RR Pant      SS Iyer   
193465  1237181       1    13     1      RR Pant      SS Iyer   
193466  1237181       1    13     2      RR Pant      SS Iyer   
193467  1237181       1    13     3      SS Iyer      RR Pant   

                 bowler  batsman_runs  extra_runs  total_runs  ...  \
0             AA Noffke             1           0           1  ...   
1             AA Noffke             1           0           1  ...   
2                Z Khan             0           0           0  ...   
3                Z Khan             1           0           1  ...   
4                Z Khan             1           0           1  ...   
...                 ...           ...         ...         ...  ...   
193463  NM Coulter-Nile             0           0           0  ...   
193464  NM Coulter-Nile             1           0           1  ...   
193465        KH Pandya             0           1           1  ...   
193466        KH Pandya             1           0           1  ...   
193467        KH Pandya             1           0           1  ...   

                        team2                  toss_winner toss_decision  \
0       Kolkata Knight Riders  Royal Challengers Bangalore         field   
1       Kolkata Knight Riders  Royal Challengers Bangalore         field   
2       Kolkata Knight Riders  Royal Challengers Bangalore         field   
3       Kolkata Knight Riders  Royal Challengers Bangalore         field   
4       Kolkata Knight Riders  Royal Challengers Bangalore         field   
...                       ...                          ...           ...   
193463         Mumbai Indians               Delhi Capitals           bat   
193464         Mumbai Indians               Delhi Capitals           bat   
193465         Mumbai Indians               Delhi Capitals           bat   
193466         Mumbai Indians               Delhi Capitals           bat   
193467         Mumbai Indians               Delhi Capitals           bat   

                       winner   result result_margin eliminator method  \
0       Kolkata Knight Riders     runs         140.0          N    NaN   
1       Kolkata Knight Riders     runs         140.0          N    NaN   
2       Kolkata Knight Riders     runs         140.0          N    NaN   
3       Kolkata Knight Riders     runs         140.0          N    NaN   
4       Kolkata Knight Riders     runs         140.0          N    NaN   
...                       ...      ...           ...        ...    ...   
193463         Mumbai Indians  wickets           5.0          N    NaN   
193464         Mumbai Indians  wickets           5.0          N    NaN   
193465         Mumbai Indians  wickets           5.0          N    NaN   
193466         Mumbai Indians  wickets           5.0          N    NaN   
193467         Mumbai Indians  wickets           5.0          N    NaN   

            umpire1      umpire2  
0         Asad Rauf  RE Koertzen  
1         Asad Rauf  RE Koertzen  
2         Asad Rauf  RE Koertzen  
3         Asad Rauf  RE Koertzen  
4         Asad Rauf  RE Koertzen  
...             ...          ...  
193463  CB Gaffaney  Nitin Menon  
193464  CB Gaffaney  Nitin Menon  
193465  CB Gaffaney  Nitin Menon  
193466  CB Gaffaney  Nitin Menon  
193467  CB Gaffaney  Nitin Menon  

[193468 rows x 35 columns]

In [64]:
df_ = df3[df3['bowling_team'].isna()]
df_not = df3[df3['bowling_team'].notna()]

In [65]:
df_[['batting_team','team1','team2']]

batting_team                        team1  \
57048              Delhi Daredevils             Delhi Daredevils   
57049              Delhi Daredevils             Delhi Daredevils   
57050              Delhi Daredevils             Delhi Daredevils   
57051              Delhi Daredevils             Delhi Daredevils   
57052              Delhi Daredevils             Delhi Daredevils   
...                             ...                          ...   
115654  Royal Challengers Bangalore  Royal Challengers Bangalore   
115655  Royal Challengers Bangalore  Royal Challengers Bangalore   
115656  Royal Challengers Bangalore  Royal Challengers Bangalore   
115657  Royal Challengers Bangalore  Royal Challengers Bangalore   
115658  Royal Challengers Bangalore  Royal Challengers Bangalore   

                   team2  
57048      Pune Warriors  
57049      Pune Warriors  
57050      Pune Warriors  
57051      Pune Warriors  
57052      Pune Warriors  
...                  ...  
115654  Rajasthan Royals  
115655  Rajasthan Royals  
115656  Rajasthan Royals  
115657  Rajasthan Royals  
115658  Rajasthan Royals  

[191 rows x 3 columns]

In [67]:
df_.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 191 entries, 57048 to 115658
Data columns (total 35 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                191 non-null    int64  
 1   inning            191 non-null    int64  
 2   over              191 non-null    int64  
 3   ball              191 non-null    int64  
 4   batsman           191 non-null    object 
 5   non_striker       191 non-null    object 
 6   bowler            191 non-null    object 
 7   batsman_runs      191 non-null    int64  
 8   extra_runs        191 non-null    int64  
 9   total_runs        191 non-null    int64  
 10  non_boundary      191 non-null    int64  
 11  is_wicket         191 non-null    int64  
 12  dismissal_kind    10 non-null     object 
 13  player_dismissed  10 non-null     object 
 14  fielder           7 non-null      object 
 15  extras_type       18 non-null     object 
 16  batting_team      191 non-null    obj

In [68]:
df_not.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 193277 entries, 0 to 193467
Data columns (total 35 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                193277 non-null  int64  
 1   inning            193277 non-null  int64  
 2   over              193277 non-null  int64  
 3   ball              193277 non-null  int64  
 4   batsman           193277 non-null  object 
 5   non_striker       193277 non-null  object 
 6   bowler            193277 non-null  object 
 7   batsman_runs      193277 non-null  int64  
 8   extra_runs        193277 non-null  int64  
 9   total_runs        193277 non-null  int64  
 10  non_boundary      193277 non-null  int64  
 11  is_wicket         193277 non-null  int64  
 12  dismissal_kind    9485 non-null    object 
 13  player_dismissed  9485 non-null    object 
 14  fielder           6777 non-null    object 
 15  extras_type       10215 non-null   object 
 16  batting_team      19

In [73]:
df_final = pd.concat([df_not, df_])
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 193468 entries, 0 to 115658
Data columns (total 35 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                193468 non-null  int64  
 1   inning            193468 non-null  int64  
 2   over              193468 non-null  int64  
 3   ball              193468 non-null  int64  
 4   batsman           193468 non-null  object 
 5   non_striker       193468 non-null  object 
 6   bowler            193468 non-null  object 
 7   batsman_runs      193468 non-null  int64  
 8   extra_runs        193468 non-null  int64  
 9   total_runs        193468 non-null  int64  
 10  non_boundary      193468 non-null  int64  
 11  is_wicket         193468 non-null  int64  
 12  dismissal_kind    9495 non-null    object 
 13  player_dismissed  9495 non-null    object 
 14  fielder           6784 non-null    object 
 15  extras_type       10233 non-null   object 
 16  batting_team      19

In [70]:
df_final["date"]= pd.to_datetime(df_final["date"])


In [71]:
df_final = df_final.sort_values(by=['date'])

In [75]:
df_final.head(5)

id  inning  over  ball      batsman  non_striker     bowler  \
0  335982       1     6     5   RT Ponting  BB McCullum  AA Noffke   
1  335982       1     6     6  BB McCullum   RT Ponting  AA Noffke   
2  335982       1     7     1  BB McCullum   RT Ponting     Z Khan   
3  335982       1     7     2  BB McCullum   RT Ponting     Z Khan   
4  335982       1     7     3   RT Ponting  BB McCullum     Z Khan   

   batsman_runs  extra_runs  total_runs  ...                  team2  \
0             1           0           1  ...  Kolkata Knight Riders   
1             1           0           1  ...  Kolkata Knight Riders   
2             0           0           0  ...  Kolkata Knight Riders   
3             1           0           1  ...  Kolkata Knight Riders   
4             1           0           1  ...  Kolkata Knight Riders   

                   toss_winner toss_decision                 winner result  \
0  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   
1  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   
2  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   
3  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   
4  Royal Challengers Bangalore         field  Kolkata Knight Riders   runs   

  result_margin eliminator method    umpire1      umpire2  
0         140.0          N    NaN  Asad Rauf  RE Koertzen  
1         140.0          N    NaN  Asad Rauf  RE Koertzen  
2         140.0          N    NaN  Asad Rauf  RE Koertzen  
3         140.0          N    NaN  Asad Rauf  RE Koertzen  
4         140.0          N    NaN  Asad Rauf  RE Koertzen  

[5 rows x 35 columns]

fielding_team is now completely processed.

In [76]:
df_final.isna().sum()

id                       0
inning                   0
over                     0
ball                     0
batsman                  0
non_striker              0
bowler                   0
batsman_runs             0
extra_runs               0
total_runs               0
non_boundary             0
is_wicket                0
dismissal_kind      183973
player_dismissed    183973
fielder             186684
extras_type         183235
batting_team             0
bowling_team             0
Partnership              0
city                  3139
date                     0
player_of_match        372
venue                    0
neutral_venue            0
team1                    0
team2                    0
toss_winner              0
toss_decision            0
winner                 372
result                 372
result_margin         3597
eliminator             372
method              190260
umpire1                  0
umpire2                  0
dtype: int64

In [80]:
df_final[df_final['winner'].isna()][['date','team1','team2']]

date                        team1             team2
121109  2015-05-17  Royal Challengers Bangalore  Delhi Daredevils
121110  2015-05-17  Royal Challengers Bangalore  Delhi Daredevils
121111  2015-05-17  Royal Challengers Bangalore  Delhi Daredevils
121112  2015-05-17  Royal Challengers Bangalore  Delhi Daredevils
121113  2015-05-17  Royal Challengers Bangalore  Delhi Daredevils
...            ...                          ...               ...
115654  2015-04-29  Royal Challengers Bangalore  Rajasthan Royals
115655  2015-04-29  Royal Challengers Bangalore  Rajasthan Royals
115656  2015-04-29  Royal Challengers Bangalore  Rajasthan Royals
115657  2015-04-29  Royal Challengers Bangalore  Rajasthan Royals
115658  2015-04-29  Royal Challengers Bangalore  Rajasthan Royals

[372 rows x 3 columns]

all these matches are no result. there are only 4 instances

In [88]:
df_final.date.unique()

array(['2008-04-18', '2008-04-19', '2008-04-20', '2008-04-21',
       '2008-04-22', '2008-04-23', '2008-04-24', '2008-04-25',
       '2008-04-26', '2008-04-27', '2008-04-28', '2008-04-29',
       '2008-04-30', '2008-05-01', '2008-05-02', '2008-05-25',
       '2008-05-03', '2008-05-04', '2008-05-05', '2008-05-06',
       '2008-05-07', '2008-05-08', '2008-05-09', '2008-05-28',
       '2008-05-10', '2008-05-11', '2008-05-12', '2008-05-13',
       '2008-05-14', '2008-05-15', '2008-05-16', '2008-05-17',
       '2008-05-18', '2008-05-19', '2008-05-20', '2008-05-21',
       '2008-05-23', '2008-05-24', '2008-05-26', '2008-05-27',
       '2008-05-30', '2008-05-31', '2008-06-01', '2009-04-18',
       '2009-04-19', '2009-04-20', '2009-04-21', '2009-04-22',
       '2009-04-23', '2009-04-24', '2009-04-25', '2009-04-26',
       '2009-04-27', '2009-04-28', '2009-04-29', '2009-04-30',
       '2009-05-01', '2009-05-02', '2009-05-03', '2009-05-04',
       '2009-05-05', '2009-05-06', '2009-05-07', '2009-

for handling cities column, excel will be used 

In [89]:
df_final.to_csv('preprocessed.csv')